_posts/2017-11-16-totw-99.md

```diff
@@ -0,0 +1,239 @@
+parameters it's a judgment call. It's usually pretty clear in practice which
+site is better. The important thing is that the specialization should not be
+hidden in client code: it should be as visible as the template and the
+user-defined types involved.
+
+### When to customize
+
+Don't customize the behavior of a class in a test. This is dangerous and
+unfortunately very common. Test source files are not immune to these dangers and
+should follow the same rules as production code. We find a lot of inappropriate
+operators in test.cc files that are written with the intent to "get `EXPECT_EQ`
+to compile" or some other pragmatic concern. Unfortunately they are still ODR
+risks (if not violations) and can make library maintenance difficult. These
+rogue definitions might even stand in the way of library maintenance as adding
+these operators upstream would break the very tests that needed them and defined
+their own. For testing there are alternatives available with a little more
+effort.
+
+Note that ADL works from the original declaration point for a type. Typedefs,
+type aliases, alias templates, and using declarations do not create types and
+have no impact on ADL. This can make it a little tricky to locate the proper
+placement for customizations, but this cannot be helped, so just do it.
+
+Don't augment the interface to types generated by protobufs. This is another
+common pitfall. You may own the `.proto` file, but that doesn't mean you own the
+C++ API it generates, and your augmentations could block improvements to the
+generated C++ API. This is an ODR risk because you can't ensure that your
+augmentations are seen whenever the generated header is included.
+
+When defining `T`, it may be tempting to define behavior for templates like
+`std::vector<T>` or `std::pair<T,T>`. Though your customizations may take
+precedence and may do what you expect, you may be conflicting with other
+expected customizations defined on the broader class template.
+
+It's possible to define some customizations for raw pointers. It may be tempting
+in some cases to supply these customizations for `T*` along with `T`. This is
+not advised. It's dangerous because the customization may conflict with the
+expected ordinary behavior of pointers (e.g. the way they're logged, swapped, or
+compared). It's best to leave pointers alone.
+
+### What to do when you're stuck
+
+Following these guidelines can be challenging. Much of the inappropriate
+overloading and specialization seen in C++ code is motivated by small set of
+root causes. Following is a partial listing with successful workarounds. If you
+run into a library API that can't be worked with, please send a note to the
+owners to see about adding the appropriate customization hooks. Common APIs should
+be usable without breaking these interface packaging guidelines.
+
+#### Testing a type with `EXPECT_EQ`, etc.
+
+*temptation:* `EXPECT_EQ` requires `operator==`, and `operator<<` or gUnit's
```

Comment:
I think the external spelling is "GUnit" but should we link to this?

(I can't find the aforementioned "PrintTo")

[GUnit](https://github.com/cpp-testing/GUnit)

Reply:
I'm not gonna worry about linking everything. PrintTo exists publicly, that's good enough for me for the moment.